### PR TITLE
a dict must not delete items during iteration

### LINKED ISSUE #2625: [Joystick Control Teleoperation] : moveit_joy node die
I try to insert my handle,but moveit_joy node still die,I can see the handle data in /joy topic.
Dictionary elements cannot be modified while traversing,but I check the code where cause the error,it seems to delete elements.
```
for name in planning_groups.keys():
            if len(planning_groups[name]) == 0:
                del planning_groups[name]
            else:
                print(name, planning_groups[name])
```
This is ouput when I run joystick_control.launch

               joy: [main]: joy_node: autorepeat_rate (40.000000 Hz) > 1/coalesce_interval (39.999999 Hz) does not make sense.
                   ~   Timing behavior is not well defined.
                 joy: [main]: Couldn't set gain on joystick force feedback: Bad file descriptor
                 joy: [main]: Opened joystick: /dev/input/js0 (Wireless Controller). deadzone_: 0.200000.
          moveit_joy: [core::RobotModel::buildModel]: Loading robot model 'engineer'...
          moveit_joy: [core::JointModel* moveit::core::RobotModel::constructJointModel]: No root/virtual joint specified in SRDF
                   ~  . Assuming fixed joint
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link back_left_wheel has visual geometry 
                   ~  but no collision geometry. Collision geometry will be left empty. Fix your URDF file by explicitly specify
                   ~  ing collision geometry.
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link back_right_wheel has visual geometry
                   ~   but no collision geometry. Collision geometry will be left empty. Fix your URDF file by explicitly specif
                   ~  ying collision geometry.
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link front_left_wheel has visual geometry
                   ~   but no collision geometry. Collision geometry will be left empty. Fix your URDF file by explicitly specif
                   ~  ying collision geometry.
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link front_right_wheel has visual geometr
                   ~  y but no collision geometry. Collision geometry will be left empty. Fix your URDF file by explicitly speci
                   ~  fying collision geometry.
          moveit_joy: [initialize]: Using empty link_prefix.
          moveit_joy: engineer_arm ['/rviz/moveit/move_marker/goal_palm']
          moveit_joy: Traceback (most recent call last):
          moveit_joy:   File "/opt/ros/noetic/lib/moveit_ros_visualization/moveit_joy.py", line 42, in <module>
          moveit_joy:     app = MoveitJoy()
          moveit_joy:   File "/opt/ros/noetic/lib/python3/dist-packages/moveit_ros_visualization/moveitjoy_module.py", line 538,
                   ~   in __init__
          moveit_joy:     self.parseSRDF()
          moveit_joy:   File "/opt/ros/noetic/lib/python3/dist-packages/moveit_ros_visualization/moveitjoy_module.py", line 516,
                   ~   in parseSRDF
          moveit_joy:     for name in planning_groups.keys():
          moveit_joy: RuntimeError: dictionary changed size during iteration
          moveit_joy: moveit_joy exited with status 1
            [rosmon]: Shutting down...
                 joy: joy exited with status 0

### PR DESCRIPTION
**Pull request by v4hn**
Fix this error:

> moveit_joy: RuntimeError: dictionary changed size during iteration

Please have a look @astroxyx , I did not even run the code.

Fixes #2625.

### CODE REVIEW

**Comment by astroxyx** [comment]
> Fix this error:
> 
> > moveit_joy: RuntimeError: dictionary changed size during iteration
> 
> Please have a look @astroxyx , I did not even run the code.
> 
> Fixes #2625.

thanks,but when I change my code like you ,it still throw error

mon launch engineer_arm_config joystick_control.launch
Loaded launch file in 0.032476s
ROS_MASTER_URI: 'http://localhost:11311'
roscore is already running.
Running as '/rosmon_1619612577756103573'
                 joy: [main]: joy_node: autorepeat_rate (40.000000 Hz) > 1/coalesce_interval (39.999999 Hz) does not make sense. Timing behavior is no
                   ~  t well defined.
                 joy: [main]: Couldn't set gain on joystick force feedback: Bad file descriptor
                 joy: [main]: Opened joystick: /dev/input/js0 (Wireless Controller). deadzone_: 0.200000.
          moveit_joy: [core::RobotModel::buildModel]: Loading robot model 'engineer'...
          moveit_joy: [core::JointModel* moveit::core::RobotModel::constructJointModel]: No root/virtual joint specified in SRDF. Assuming fixed joint
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link back_left_wheel has visual geometry but no collision geome
                   ~  try. Collision geometry will be left empty. Fix your URDF file by explicitly specifying collision geometry.
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link back_right_wheel has visual geometry but no collision geom
                   ~  etry. Collision geometry will be left empty. Fix your URDF file by explicitly specifying collision geometry.
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link front_left_wheel has visual geometry but no collision geom
                   ~  etry. Collision geometry will be left empty. Fix your URDF file by explicitly specifying collision geometry.
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link front_right_wheel has visual geometry but no collision geo
                   ~  metry. Collision geometry will be left empty. Fix your URDF file by explicitly specifying collision geometry.
          moveit_joy: [initialize]: Using empty link_prefix.
          moveit_joy: Traceback (most recent call last):
          moveit_joy:   File "/opt/ros/noetic/lib/moveit_ros_visualization/moveit_joy.py", line 42, in <module>
          moveit_joy:     app = MoveitJoy()
          moveit_joy:   File "/opt/ros/noetic/lib/python3/dist-packages/moveit_ros_visualization/moveitjoy_module.py", line 538, in __init__
          moveit_joy:     self.parseSRDF()
          moveit_joy:   File "/opt/ros/noetic/lib/python3/dist-packages/moveit_ros_visualization/moveitjoy_module.py", line 514, in parseSRDF
          moveit_joy:     if len(self.planning_groups[g]) > 0:
          moveit_joy: AttributeError: 'MoveitJoy' object has no attribute 'planning_groups'
          moveit_joy: moveit_joy exited with status 1
            [rosmon]: Shutting down...
                 joy: joy exited with status 0

**Comment by v4hn** [comment]
Sorry, there was a typo in the variable.
I can't test that code right now, so I write the patch without verification...

**Comment by v4hn** [comment]
Could you please test once more @astroxyx and maybe provide a fixup yourself if it should still fail?
If you don't this PR is likely to rot because it touches an edge-case of a minor script few people work with.

**Comment by astroxyx** [comment]
> Could you please test once more @astroxyx and maybe provide a fixup yourself if it should still fail?
> If you don't, this PR is likely to rot because it touches an edge-case of a minor script few people work with.

thank you very much,I will try to fix it.

**Comment by astroxyx** [comment]
I fix the code by change some Dictionaries into list .Now it seems to waiting for the /rviz/moveit/move_marker/goal_palm topic,but I can see this topic already exists.
This is output:

                 joy: [main]: joy_node: autorepeat_rate (40.000000 Hz) > 1/coalesce_interval (39.
                   ~  999999 Hz) does not make sense. Timing behavior is not well defined.
                 joy: [main]: Opened joystick: /dev/input/js0 (Wireless Controller). deadzone_: 0
                   ~  .200000.
                 joy: [main]: Couldn't set gain on joystick force feedback: Bad file descriptor
          moveit_joy: [core::RobotModel::buildModel]: Loading robot model 'engineer'...
          moveit_joy: [core::JointModel* moveit::core::RobotModel::constructJointModel]: No root/
                   ~  virtual joint specified in SRDF. Assuming fixed joint
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link left_
                   ~  back_wheel has visual geometry but no collision geometry. Collision geometr
                   ~  y will be left empty. Fix your URDF file by explicitly specifying collision
                   ~   geometry.
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link left_
                   ~  front_wheel has visual geometry but no collision geometry. Collision geomet
                   ~  ry will be left empty. Fix your URDF file by explicitly specifying collisio
                   ~  n geometry.
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link right
                   ~  _back_wheel has visual geometry but no collision geometry. Collision geomet
                   ~  ry will be left empty. Fix your URDF file by explicitly specifying collisio
                   ~  n geometry.
          moveit_joy: [core::LinkModel* moveit::core::RobotModel::constructLinkModel]: Link right
                   ~  _front_wheel has visual geometry but no collision geometry. Collision geome
                   ~  try will be left empty. Fix your URDF file by explicitly specifying collisi
                   ~  on geometry.
          moveit_joy: [initialize]: Using empty link_prefix.
          moveit_joy: engineer_arm ['/rviz/moveit/move_marker/goal_palm']
          moveit_joy: [MoveitJoy.updatePlanningGroup]: Changed planning group to engineer_arm
          moveit_joy: [MoveitJoy.updatePoseTopic]: Changed controlled end effector to palm
          moveit_joy: [MoveitJoy.updatePlanningGroup]: Changed planning group to engineer_arm
          moveit_joy: [MoveitJoy.waitForInitialPose]: Waiting for pose topic of 'goal_palm' to be
                   ~   initialized
          moveit_joy: [MoveitJoy.waitForInitialPose]: Waiting for pose topic of 'goal_palm' to be
                   ~   initialized
          moveit_joy: [MoveitJoy.updatePlanningGroup]: Changed planning group to engineer_arm
          moveit_joy: [MoveitJoy.waitForInitialPose]: Waiting for pose topic of 'goal_palm' to be
                   ~   initialized
          moveit_joy: [MoveitJoy.waitForInitialPose]: Waiting for pose topic of 'goal_palm' to be
                   ~   initialized
          moveit_joy: [MoveitJoy.updatePlanningGroup]: Changed planning group to engineer_arm
          moveit_joy: [MoveitJoy.waitForInitialPose]: Waiting for pose topic of 'goal_palm' to be
                   ~   initialized

File: moveit_ros/visualization/src/moveit_ros_visualization/moveitjoy_module.py [hunk 1]
```diff
@@ -523,15 +523,13 @@ def parseSRDF(self):
...
+                planning_groups[g] = [
+                    "/rviz/moveit/move_marker/goal_" + l
+                    for l in self.planning_groups_tips[g]
+                ]
         for name in planning_groups.keys():
-            if len(planning_groups[name]) == 0:
-                del planning_groups[name]
-            else:
-                print(name, planning_groups[name])
+            print(name, planning_groups[name])
```
Comment (jliukkonen) [inline moveit_ros/visualization/src/moveit_ros_visualization/moveitjoy_module.py:532]: Nit: I bet that fixes the error nicely but "while there", why not write it like this:
```
for group_name, planning_groups in planning_groups:
    print(group_name, planning_groups)
```
Either way, looks good to me.

**Review by jliukkonen — APPROVED** [review]

**Comment by v4hn** [comment]
Sorry @astroxyx , as you noticed nobody has the spare cycles to look at this particular code.
If you are still interested in it and wrote some patches to get it to work on your side, please create pull-requests.
